File: TileCalorimeter/TileMonitoring/src/TileRawChannelFlxMonitorAlgorithm.cxx
/*
  Copyright (C) 2002-2023 CERN for the benefit of the ATLAS collaboration
*/

// Tile includes
#include "TileRawChannelFlxMonitorAlgorithm.h"
#include "TileIdentifier/TileHWID.h"
#include "TileCalibBlobObjs/TileCalibUtils.h"

// Athena includes
#include "StoreGate/ReadHandle.h"

#include <cstring>


StatusCode TileRawChannelFlxMonitorAlgorithm::initialize() {

  ATH_CHECK( AthMonitorAlgorithm::initialize() );

  ATH_MSG_INFO("in initialize()");

  ATH_CHECK( detStore()->retrieve(m_tileHWID) );
  ATH_CHECK( m_rawChannelContainerKeyLegacy.initialize() );
  ATH_CHECK( m_rawChannelContainerKeyFlx.initialize() );

  std::ostringstream os;

  if ( m_fragIDsToCompare.size() != 0) {
    std::sort(m_fragIDsToCompare.begin(), m_fragIDsToCompare.end());
    for (int fragID : m_fragIDsToCompare) {
      unsigned int ros    = fragID >> 8;
      unsigned int drawer = fragID & 0xFF;
      std::string module = TileCalibUtils::getDrawerString(ros, drawer);
      os << " " << module << "/0x" << std::hex << fragID << std::dec;
    }
  } else {
    os << "NONE";
  }

  ATH_MSG_INFO("Monitored modules/frag ID:" << os.str());

  return StatusCode::SUCCESS;
}


StatusCode TileRawChannelFlxMonitorAlgorithm::fillHistograms( const EventContext& ctx ) const {

  // In case you want to measure the execution time
  auto timer = Monitored::Timer("TIME_execute");

  SG::ReadHandle<TileRawChannelContainer> rawChannelContainerLegacy(m_rawChannelContainerKeyLegacy, ctx);
  ATH_CHECK( rawChannelContainerLegacy.isValid() );

  SG::ReadHandle<TileRawChannelContainer> rawChannelContainerFlx(m_rawChannelContainerKeyFlx, ctx);
  ATH_CHECK( rawChannelContainerFlx.isValid() );

  std::array<std::string, 2> gainName{"_LG", "_HG"};

  float amplitude[TileCalibUtils::MAX_CHAN][TileCalibUtils::MAX_GAIN];
  float amplitudeFlx[TileCalibUtils::MAX_CHAN][TileCalibUtils::MAX_GAIN];
  int found[TileCalibUtils::MAX_CHAN][TileCalibUtils::MAX_GAIN];

  const TileFragHash& hashFunc = rawChannelContainerFlx->hashFunc();

  for (int fragID : m_fragIDsToCompare) {

    memset(found,0,sizeof(found));

    IdentifierHash hash = static_cast<IdentifierHash>(hashFunc(fragID));
    unsigned int drawer = (fragID & 0x3F);
    unsigned int ros = fragID >> 8;
    std::string module = TileCalibUtils::getDrawerString(ros, drawer);
    std::array<std::string, 2> moduleName{module+gainName[0], module+gainName[1]};

    // Legacy amplitudes
    const TileRawChannelCollection* rawChannelCollectionLegacy = rawChannelContainerLegacy->indexFindPtr(hash);
    for (const TileRawChannel* rawChannel : *rawChannelCollectionLegacy) {

      HWIdentifier adcId = rawChannel->adc_HWID();
      int channel = m_tileHWID->channel(adcId);
      int gain = m_tileHWID->adc(adcId);

      found[channel][gain] |= 1;
      amplitude[channel][gain] = rawChannel->amplitude();

      std::string channelName = moduleName[gain] + "_channel";
      auto monitoredChannel = Monitored::Scalar<float>(channelName, channel);

      std::string amplitudeName = moduleName[gain] + "_amplitude";
      auto monitoredAmplitude = Monitored::Scalar<float>(amplitudeName, rawChannel->amplitude());
      fill("TileRawChannelAmpLegacy", monitoredChannel, monitoredAmplitude);
    }

    // FELIX amplitudes
    const TileRawChannelCollection* rawChannelCollectionFlx = rawChannelContainerFlx->indexFindPtr(hash);
    for (const TileRawChannel* rawChannel : *rawChannelCollectionFlx) {

      HWIdentifier adcId = rawChannel->adc_HWID();
      int channel = m_tileHWID->channel(adcId);
      int gain = m_tileHWID->adc(adcId);

      found[channel][gain] |= 2;
      amplitudeFlx[channel][gain] = rawChannel->amplitude();

      std::string channelName = moduleName[gain] + "_channel";
      auto monitoredChannel = Monitored::Scalar<float>(channelName, channel);

      std::string amplitudeName = moduleName[gain] + "_amplitude";
      auto monitoredAmplitude = Monitored::Scalar<float>(amplitudeName, rawChannel->amplitude());
      fill("TileRawChannelAmpFlx", monitoredChannel, monitoredAmplitude);
    }

    // Compare amplitude and amplitudeFlx arrays and put results into histograms
    for(unsigned int gain = 0; gain<TileCalibUtils::MAX_GAIN; ++gain) {

      std::string channelName = moduleName[gain] + "_channel";
      std::string amplitudeName  = moduleName[gain] + "_amplitude";
      std::string amplitudeDiffName = moduleName[gain] + "_amplitude_diff";

      auto monitoredChannel = Monitored::Scalar<float>(channelName, 0.0F);
      auto monitoredAmplitude = Monitored::Scalar<float>(amplitudeName, 0.0F);
      auto monitoredAmplitudeDiff = Monitored::Scalar<float>(amplitudeDiffName, 0.0F);

      for(unsigned int channel = 0; channel<TileCalibUtils::MAX_CHAN; ++channel) {

        if (found[channel][gain] == 3) {

          monitoredChannel = channel;
          monitoredAmplitude = amplitude[channel][gain];
          monitoredAmplitudeDiff = amplitudeFlx[channel][gain] - amplitude[channel][gain] * m_felixScale;

          fill("TileRawChannelAmpDiff", monitoredChannel, monitoredAmplitudeDiff);
          fill("TileRawChannelAmpDiffVsLegacy", monitoredAmplitude, monitoredAmplitudeDiff);
        }
      }
    }
  }

  fill("TileRawChannelFlxMonExecuteTime", timer);

  return StatusCode::SUCCESS;
}

File: TileCalorimeter/TileMonitoring/python/TileRawChannelFlxMonitorAlgorithm.py
#
#  Copyright (C) 2002-2024 CERN for the benefit of the ATLAS collaboration
#

'''
@file TileDigitsFlxMonitorAlgorithm.py
@brief Python configuration of TileDigitsFlxMonitorAlgorithm algorithm for the Run III
'''

def TileRawChannelFlxMonitoringConfig(flags, fragIDs=[0x201, 0x402], **kwargs):
    '''Function to configures TileRawChannelFlxMonAlg algorithms in the monitoring system.'''

    kwargs.setdefault('TileRawChannelContainerLegacy', 'TileRawChannelFit')
    kwargs.setdefault('TileRawChannelContainerFlx', 'TileRawChannelFlxFit')
    kwargs.setdefault('FelixScale', 4)

    felixScale = kwargs['FelixScale']
    
    from AthenaConfiguration.ComponentAccumulator import ComponentAccumulator
    result = ComponentAccumulator()

    from TileConditions.TileCablingSvcConfig import TileCablingSvcCfg
    result.merge( TileCablingSvcCfg(flags) )

    from AthenaMonitoring import AthMonitorCfgHelper
    helper = AthMonitorCfgHelper(flags, 'TileRawChannelFlxMonitoring')

    from AthenaConfiguration.ComponentFactory import CompFactory
    TileRawChannelFlxMonitorAlgorithm = CompFactory.TileRawChannelFlxMonitorAlgorithm
    tileRawChannelFlxMonAlg = helper.addAlgorithm(TileRawChannelFlxMonitorAlgorithm, 'TileRawChannelFlxMonAlg')
    tileRawChannelFlxMonAlg.TriggerChain = ''

    for k, v in kwargs.items():
        setattr(tileRawChannelFlxMonAlg, k, v)

    # Configure histogram with TileRawChannelFlxMonAlg algorithm execution time
    executeTimeGroup = helper.addGroup(tileRawChannelFlxMonAlg, 'TileRawChannelFlxMonExecuteTime', 'Tile')
    executeTimeGroup.defineHistogram('TIME_execute', path = 'RawChannel', type='TH1F',
                                     title = 'Time for execute TileRawChannelFlxMonAlg algorithm;time [#mus]',
                                     xbins = 300, xmin = 0, xmax = 300000)

    runNumber = flags.Input.RunNumbers[0]
    from TileCalibBlobObjs.Classes import TileCalibUtils as Tile

    modules = []
    if fragIDs:
        for fragID in fragIDs:
            ros = fragID >> 8
            drawer = fragID & 0x3F
            modules += [Tile.getDrawerString(ros, drawer)]
    else:
        for ros in range(1, Tile.MAX_ROS):
            for drawer in range(0, Tile.MAX_DRAWER):
                modules += [Tile.getDrawerString(ros, drawer)]

    channelLegacyGroup = helper.addGroup(tileRawChannelFlxMonAlg, 'TileRawChannelAmpLegacy', 'Tile/Legacy/RawChannel')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {str(runNumber)} {moduleName} {gainName}: Amplitude (Legacy);Channel;Amplitude'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_amplitude;{moduleName}_{gainName}_amplitude_legacy'
            path = moduleName
            channelLegacyGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                               xbins = 48, xmin = -0.5, xmax = 47.5)

    channelFelixGroup = helper.addGroup(tileRawChannelFlxMonAlg, 'TileRawChannelAmpFlx', 'Tile/Felix/RawChannel')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {str(runNumber)} {moduleName} {gainName}: Amplitude (FELIX);Channel;Amplitude'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_amplitude;{moduleName}_{gainName}_amplitude_felix'
            path = moduleName
            channelFelixGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                              xbins = 48, xmin = -0.5, xmax = 47.5)

    channelCompareGroup = helper.addGroup(tileRawChannelFlxMonAlg, 'TileRawChannelAmpDiff', 'Tile/Compare/RawChannel')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {str(runNumber)} {moduleName} {gainName}: Amplitude difference (FELIX-Legacy*{felixScale});Channel;Amplitude difference [ADC]'
            name = f'{moduleName}_{gainName}_channel,{moduleName}_{gainName}_amplitude_diff;{moduleName}_{gainName}_amplitude_diff'
            path = moduleName
            channelCompareGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                                xbins = 48, xmin = -0.5, xmax = 47.5)
    
    channelCompareVsLegacyGroup = helper.addGroup(tileRawChannelFlxMonAlg, 'TileRawChannelAmpDiffVsLegacy', 'Tile/Compare/RawChannel')
    for moduleName in modules:
        for gainName in ['HG', 'LG']:
            title = f'Run {str(runNumber)} {moduleName} {gainName}: Amplitude difference (FELIX-Legacy*{felixScale});Amplitude (Legacy) [ADC];Amplitude difference [ADC]'
            name = f'{moduleName}_{gainName}_amplitude,{moduleName}_{gainName}_amplitude_diff;{moduleName}_{gainName}_amplitude_diff_vs_legacy'
            path = moduleName
            channelCompareVsLegacyGroup.defineHistogram(name, title = title, path = path, type = 'TProfile',
                                                        xbins = 1024, xmin = -0.5, xmax = 1023.5)

    accumalator = helper.result()
    result.merge(accumalator)
    return result



if __name__=='__main__':

    # Setup logs
    from AthenaCommon.Logging import log
    from AthenaCommon.Constants import INFO
    log.setLevel(INFO)

    # Set the Athena configuration flags
    from AthenaConfiguration.AllConfigFlags import initConfigFlags
    from AthenaConfiguration.TestDefaults import defaultTestFiles
    flags = initConfigFlags()
    flags.Input.Files = defaultTestFiles.RAW_RUN2
    flags.Output.HISTFileName = 'TileRawChannelFlxMonitorOutput.root'
    flags.DQ.useTrigger = False
    flags.DQ.enableLumiAccess = False
    flags.Exec.MaxEvents = 3
    flags.fillFromArgs()
    flags.lock()

    # Initialize configuration object, add accumulator, merge, and run.
    from AthenaConfiguration.MainServicesConfig import MainServicesCfg
    cfg = MainServicesCfg(flags)

    from TileByteStream.TileByteStreamConfig import TileRawDataReadingCfg
    cfg.merge( TileRawDataReadingCfg(flags, readMuRcv=False) )

    cfg.merge(TileRawChannelFlxMonitoringConfig(flags,
                                                TileRawChannelContainerLegacy='TileRawChannelFit',
                                                TileRawChannelContainerFlx='TileRawChannelFlxFit',
                                                fillHistogramsForDSP=True))

    flags.dump()

    cfg.store( open('TileRawChannelMonitorAlgorithm.pkl','wb') )

    sc = cfg.run()

    import sys
    # Success should be 0
    sys.exit(not sc.isSuccess())

